io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/writer/internal/StreamGmlWriter.java

```diff
@@ -1151,6 +1152,45 @@ public URI getLocation() {
 
 			}
 
+			// get additional schema location from all the imports
+			// for all the instances
+			ResourceIterator<Instance> resourceIterator = instances.iterator();
+			DefaultInstanceCollection itInstancenew = new DefaultInstanceCollection(
+					new ArrayList<Instance>());
+			try {
+				Set<String> schemasToBeAdded = new LinkedHashSet<String>();
+				while (resourceIterator.hasNext() && !progress.isCanceled()) {
+					Instance instance = resourceIterator.next();
+					itInstancenew.add(instance);
```

Comment:
Like this every instance will be kept in memory, which will be a problem if there is a lot of data.

Reply:
If you refer to schemasToBeAdded from what I've seen till now this list could have maximum couple of elements. 
If you refer to the resourceIterator, I'm closing it after using it.
I would love to use the instances.iterator() twice also not to have the itInstancenew, but somehow for the tests accessing it second time returns always null. Instead from the HS works as it sholud.
Do you have any better idea?
Thank you

Reply:
I refer to all instances being added to a collection during the process and thus remaining in memory.

If the instances need to be traversed multiple times, then the approach as before iterating two times is the correct one, but the implementation needs to state that it can't process the instances in one iteration (see `isPassthrough`method).
However, as mentioned in the other discussion if we do the two iterations here, I would want that behavior to not be the default and enabled with a configuration parameter, to by default keep the possibility to stream the results:

> If it is certain that only the feature type namespace is required to be covered and not any others that appear in the file from the features' elements on, then I would make the behavior as you implemented it now an optional behavior, to still keep the possibility to stream instances using the GML writer.